core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -59,25 +62,32 @@
   private final Multimap<Integer, Move> moves = Multimaps.newListMultimap(Maps.newHashMap(), Lists::newArrayList);
   private int lastColumnId;
   private boolean allowIncompatibleChanges = false;
-
+  private Set<String> identifierNames;
```

Comment:
I would prefer not to track these in the update by name since it makes the code more complex. I think the only missing part is tracking what fields have been added, for which I think we could keep a map from added name to new ID that is used if the current schema doesn't have the field.

I understand this has taken a while to get in, so we can do that in a follow-up if you'd like.